File: tests/drivers/node/v5_8/parallel_edge_import.js
var neo4j = require('neo4j-driver');
var driver = neo4j.driver("bolt://localhost:7687",
                          neo4j.auth.basic("", ""),
                          { encrypted: 'ENCRYPTION_OFF', maxTransactionRetryTime: 30000 });
var session = driver.session();

var node_count = 50;

function die() {
  session.close();
  driver.close();
  process.exit(1);
}

async function run_query(query, callback) {
  var run = session.run(query, {});
  run.then(callback).catch(function (error) {
    console.log(error);
    die();
  });
}

async function run_query_with_sess(sess, query, callback) {
  const result = await sess.executeWrite(async tx => {
    return await tx.run(query);
  });
  console.log("Transaction got through!");
  sess.close();
  return result;
}

async function run_conflicting_queries(query, number, callback) {
  const promises = [];
  for (let i = 0; i < number; i++) {
    var sess = driver.session();
    promises.push(run_query_with_sess(sess, query));
  }
  await Promise.all(promises).then(callback);
}

async function assert_node_count(expectedCount) {
  try {
    const run = await session.run("MATCH (n) RETURN count(n) AS cnt", {});
    const result = run.records[0].get("cnt").toNumber();
    if (result !== expectedCount) {
      console.log("Count result is not correct! (" + result + ")");
      die();
    } else {
      console.log("Count result is correct! (" + result + ")");
    }
  } catch (error) {
    console.log(error);
    die();
  } finally {
    session.close();
    driver.close();
  }
}

run_query("MATCH (n) DETACH DELETE n;", function (result) {
  console.log("Database cleared.");
  run_query("MERGE (root:Root);", function (result) {
    console.log("Root created.");
    run_conflicting_queries("MATCH (root:Root) CREATE (n:Node) CREATE (n)-[:TROUBLING_EDGE]->(root)", node_count - 1, function (result) {
      assert_node_count(node_count);
    });
  });
});

File: tests/drivers/python/v5_8/parallel_edge_import.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright 2021 Memgraph Ltd.
#
# Use of this software is governed by the Business Source License
# included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
# License, and you may not use this file except in compliance with the Business Source License.
#
# As of the Change Date specified in that file, in accordance with
# the Business Source License, use of this software will be governed
# by the Apache License, Version 2.0, included in the file
# licenses/APL.txt.

import sys
import threading

from neo4j import GraphDatabase, basic_auth

driver = GraphDatabase.driver("bolt://localhost:7687", auth=("", ""), encrypted=False)
session = driver.session()

node_count = 100


def add_edge_tx(tx):
    result = tx.run("MATCH (root:Root) CREATE (n:Node) CREATE (n)-[:TROUBLING_EDGE]->(root) RETURN root").single()
    pass


def add_edge():
    sess = driver.session()
    sess.execute_write(add_edge_tx)
    print("Transaction got through!")
    sess.close()


session.run("MATCH (n) DETACH DELETE n").consume()
print("Database cleared.")

session.run("MERGE (root:Root);").consume()
print("Root created.")

threads = [threading.Thread(target=add_edge) for x in range(node_count - 1)]
[x.start() for x in threads]
[x.join() for x in threads]

count = session.run("MATCH (n) RETURN count(n) AS cnt").single()["cnt"]
assert count == node_count

session.close()
driver.close()

print("All ok!")
